src/binary/ibc-setup/commands/keys-list.ts

```diff
@@ -0,0 +1,93 @@
+import fs from 'fs';
+import os from 'os';
+import path from 'path';
+import readline from 'readline';
+
+import { registryFile } from '../../constants';
+import { deriveAddress } from '../utils/derive-address';
+import { getDefaultHomePath } from '../utils/get-default-home-path';
+import { loadAndValidateRegistry } from '../utils/load-and-validate-registry';
+import { resolveRequiredOption } from '../utils/resolve-required-option';
+
+type Flags = {
+  readonly interactive: boolean;
+  readonly home?: string;
+  readonly keyFile?: string;
+  readonly mnemonic?: string;
+};
+
+export type Options = {
+  readonly home: string;
+  readonly mnemonic: string;
+};
+
+export async function readMnemonicFromStdin(interactive: Flags['interactive']) {
+  if (!interactive) {
+    return undefined;
+  }
+
+  const readlineInterface = readline.createInterface({
+    input: process.stdin,
+    output: process.stdout,
+  });
+
+  const mnemonic = await new Promise<string>((resolve) => {
+    readlineInterface.question('enter mnemonic phrase: ', (stdin) => {
+      readlineInterface.close();
+      resolve(stdin);
+    });
+  });
+
+  return mnemonic;
+}
+
+function readMnemonicFromFile(keyFile: Flags['keyFile']) {
+  if (!keyFile) {
+    return undefined;
+  }
+
+  return () => {
+    return fs.readFileSync(keyFile, 'utf-8').trim();
+  };
+}
+
+export async function keysList(flags: Flags) {
+  const options: Options = {
+    home: resolveRequiredOption('home')(
+      flags.home,
+      process.env.RELAYER_HOME,
+      getDefaultHomePath
+    ),
+    mnemonic: resolveRequiredOption('mnemonic')(
```

Comment:
This is a very interesting way to handle this. It surprised me at first and I was like "no, this doesn't make sense". But seeing how each of these (readMnemonicFromStdin and readMnemonicFromFile) returns undefined if the flag was not set (becomes a no-op), this is a great way to handle order of operations:

* if `-i`, then read from stdin
* if `--mnemonic`, use it
* if `--keyfile`, load from file
* if `RELAYER_MNEMONIC` set, use that

My only open question now is how the `app.yaml` flag values are loaded?

Reply:
After more reflection I think the proper handling would be to check each flag one after another. Something like:

```ts
let mnemonic = await readMnemonicFromStdin(flags.mneomic);
if (!mnemonic) {
  const mnemonic = resolveOption(flags.mnemonic, app.mnemonic, process.env.RELAY_MNEMONIC);
}
if (!mnemonic) {
  const keyfile = resolveOption(flags.keyFile, app.keyFile, process.env.RELAY_KEY_FILE);
  if (keyfile) {
    mnemonic = readMnemonicFromFile(keyfile);
  }
}
if (!mnemonic) {
  throw new Error('You must provide the mnemonic via `interactive`, `mnemonic` or `key-file`');
```

Reply:
Yes, this is not so clear as using the same loop, but the logic is different. We can wrap this in one helper function that takes flags and app config. and just use it.

The `resolveOption` and `requireResolveOption` can be used everywhere else, but not for doing the checks over multiple possible flags

Reply:
I see your point, but I wanted to give it one more try: eb4a75e

Here, we prepare the `keyFile` before resolving the mnemonic:
```ts
const home = resolveRequiredOption('home')(...)
const app = loadAndValidateApp(home);
const keyFile = resolveOption(flags.keyFile, app?.keyFile, process.env.KEY_FILE);

const mnemonic = await resolveMnemonicOption({ app, keyFile, interactive: flags.interactive, mnemonic: flags.mnemonic });
```

In **resolveMnemonicOption**, the `resolveRequiredOption` is used to determine the order.

**Note: ** I introduced `null`s to indicate values explicitly set to "nothing" by us. It may look like a mishmash at first, but it does make sense when you think of `undefined` as a value that comes from the outside (flags, configs) and a `null` that is explicitly set in the program. I'm not 100% sure about this, but it gives me more confidence.

What do you think of such an approach (mnemonic and nulls)?